tests/test_infraction.py

```diff
@@ -0,0 +1,180 @@
+
+
+def access_control_error_message(address, role=None):
+    role = role or b"\x00" * 32
+    return f"account={address}, neededRole={role}"
+
+
+@pytest.fixture(scope="module")
+def nodes(accounts):
+    return sorted(accounts[:MAX_DKG_SIZE], key=lambda x: x.address.lower())
+
+
+@pytest.fixture(scope="module")
+def initiator(accounts):
+    initiator_index = MAX_DKG_SIZE + 1
+    assert len(accounts) >= initiator_index
+    return accounts[initiator_index]
+
+
+@pytest.fixture(scope="module")
+def deployer(accounts):
+    deployer_index = MAX_DKG_SIZE + 2
+    assert len(accounts) >= deployer_index
+    return accounts[deployer_index]
+
+
+@pytest.fixture(scope="module")
+def treasury(accounts):
+    treasury_index = MAX_DKG_SIZE + 3
+    assert len(accounts) >= treasury_index
+    return accounts[treasury_index]
+
+
+@pytest.fixture()
+def application(project, deployer, nodes):
+    contract = project.ChildApplicationForCoordinatorMock.deploy(sender=deployer)
+    for n in nodes:
+        contract.updateOperator(n, n, sender=deployer)
+        contract.updateAuthorization(n, 42, sender=deployer)
+    return contract
+
+
+@pytest.fixture()
+def erc20(project, initiator):
+    token = project.TestToken.deploy(ERC20_SUPPLY, sender=initiator)
+    return token
+
+
+@pytest.fixture()
+def coordinator(project, deployer, application, erc20, initiator, oz_dependency):
+    admin = deployer
+    contract = project.Coordinator.deploy(
+        application.address,
+        erc20.address,
+        FEE_RATE,
+        sender=deployer,
+    )
+
+    encoded_initializer_function = contract.initialize.encode_input(TIMEOUT, MAX_DKG_SIZE, admin)
+    proxy = oz_dependency.TransparentUpgradeableProxy.deploy(
+        contract.address,
+        deployer,
+        encoded_initializer_function,
+        sender=deployer,
+    )
+    proxy_contract = project.Coordinator.at(proxy.address)
+
+    proxy_contract.grantRole(contract.INITIATOR_ROLE(), initiator, sender=admin)
+    return proxy_contract
+
+
+@pytest.fixture()
+def global_allow_list(project, deployer, coordinator):
+    contract = project.GlobalAllowList.deploy(coordinator.address, sender=deployer)
+    return contract
+
+
+@pytest.fixture
+def infraction_collector(project, deployer, coordinator, application, oz_dependency):
+    contract = project.InfractionCollector.deploy(
+        coordinator.address, application.address, sender=deployer
+    )
+    proxy = oz_dependency.TransparentUpgradeableProxy.deploy(
+        contract.address,
+        deployer,
+        b"",
+        sender=deployer,
+    )
+    proxy_contract = project.InfractionCollector.at(proxy.address)
+    return proxy_contract
+
+
+def test_no_infractions(
+    erc20, nodes, initiator, global_allow_list, infraction_collector, coordinator
+):
+    for node in nodes:
+        public_key = gen_public_key()
+        coordinator.setProviderPublicKey(public_key, sender=node)
+        cost = coordinator.getRitualInitiationCost(nodes, DURATION)
```

Comment:
Should this line for `cost` be outside of the `for` loop? It isn't node-specific.

Same comment for other `cost = ...` lines in subsequent tests.